File: 739-daily-temperatures/daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {   
        Stack<Integer> stack = new Stack<>();
        int[] result = new int[temperatures.length];

        for(int i = 0; i < temperatures.length; i++){
            //when the stack isnt empty and while the temp[i] > other prev values
            while(!stack.isEmpty() && temperatures[i] > temperatures[stack.peek()]){

                //pop the index since we found a greater element
                int index = stack.pop();

                //calculate the distance from each other
                result[index] = i - index;
            }
            
            //else push this onto stack
            stack.push(i);
        }

        return result;
        
    }
}

File: 739-daily-temperatures/daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack = []
        result = [0] * len(temperatures)

        for i in range(len(temperatures)):

            # while stack isnt empty and temp is warmer than prev
            while stack and temperatures[i] > temperatures[stack[-1]]:
                # pop prev warmest
                index = stack.pop()
                result[index] = i - index 

            stack.append(i)
        
        return result